7660 Avoid CI stability checks timing out

DanielRyanSmith opened the pull request:
addresses issue #7660

This is implemented based on [the proposed solution](https://docs.google.com/document/d/1dAlCSHUQldtgWDDTrGJR-ksm19FZZ3k8ppqc5-kSwIk/edit#) described by @stephenmcgruer.

Changes:
- Adds a new `--repeat-max-time` flag that will set how many minutes wptrunner should attempt to run the iterations of the test suite for the stability checks. Defaults to 100 minutes.
- When it is anticipated that an additional iteration run will eclipse this max time, the iterations of the test suite will terminate early.
- The test results will be evaluated based on the number of iterations run rather than the number attempted (10 by default). **As long as at least 2 or more iterations were completed.
- `wptrunner.run_tests` has had some refactoring to trim down the size of the function to be more readable.

DanielRyanSmith committed:
allow stability checks to avoid TC timeout by checking times in between repeat runs

DanielRyanSmith committed:
fix flake8 issue

DanielRyanSmith committed:
remove empty flag to trigger stability checks

DanielRyanSmith committed:
some commenting explaining how iterations are being tracked

DanielRyanSmith committed:
add --repeat-max-time flag

DanielRyanSmith committed:
better descriptors for announcing results

If the repeat runs stop early to avoid, timeout, a message will be written in the results explaining so.

DanielRyanSmith committed:
cast max_time to timedelta

DanielRyanSmith committed:
correct syntax for kwargs

DanielRyanSmith committed:
specify kwargs source for run_test_iteration

DanielRyanSmith committed:
remove empty css flags tag to trigger stability checks

DanielRyanSmith committed:
Add clause to handle an ineffective number of iterations completing

DanielRyanSmith commented on css/CSS2/borders/border-001.xht:
```diff
@@ -9,7 +9,6 @@
         <link rel="help" href="http://www.w3.org/TR/css3-background/#borders"/>
         <link rel="help" href="http://www.w3.org/TR/css3-background/#the-border-shorthands"/>
         <link rel="match" href="border-001-ref.xht"/>
-        <meta name="flags" content=""/>
```
Used to trigger the stability checks to actually run. Can be removed.

DanielRyanSmith committed:
replace unassociated change used to trigger stability checks

stephenmcgruer commented on tools/wptrunner/wptrunner/wptrunner.py:
```diff
@@ -235,143 +374,78 @@ def run_tests(config, test_paths, product, **kwargs):
+                # determine the longest test suite runtime seen
+                longest_iteration_time = max(
+                    longest_iteration_time,
+                    datetime.now() - iteration_start)
```
Vague musing; I wonder if we should 'pad' the iteration time by e.g. +10%, to avoid the case where we estimate we will complete but actually narrowly avoid it.

DanielRyanSmith replied:
I do think this is a good idea and not terribly hard to implement - I've added the change.

stephenmcgruer reviewed (commented):
This generally LGTM as an implementation of my old design. If ok, I would rather leave it to current WPT maintainers to do a 'real' review of the code, as they will be the ones maintaining it long term.

Side-note: there's a bunch of changes here that seem to be the result of some autoformatting happening in your editor. It would be great to revert those changes, as (a) they're inconsistent with WPT styling elsewhere I think (though we don't have a real style guide afaik...) but more importantly (b) they make it much harder to spot which deltas are new and which are just reformats.

DanielRyanSmith committed:
Implement changes suggested by @stephenmcgruer

DanielRyanSmith committed:
Add only necessary formatting changes to stability

DanielRyanSmith committed:
Merge branch 'master' into 7660-stability-timeout

DanielRyanSmith commented on tools/wptrunner/wptrunner/stability.py:
```diff
@@ -270,6 +271,13 @@ def run_step(logger, iterations, restart_after_iteration, kwargs_extras, **kwarg
+    # hitting the timeout. If so, the actual number of iterations run
+    # should be used to process the results. The number here will
+    # be set to the actual value for later reference if it changes.
+    kwargs["avoided_timeout"] = {"did_avoid": False,
```
I agree with this - definitely a clearer naming convention

DanielRyanSmith commented on tools/wptrunner/wptrunner/stability.py:
```diff
@@ -335,8 +361,9 @@ def write_summary(logger, step_results, final_result):
-def check_stability(logger, repeat_loop=10, repeat_restart=5, chaos_mode=True, max_time=None,
-                    output_results=True, **kwargs):
+
+def check_stability(logger, repeat_loop=10, repeat_restart=10, chaos_mode=True,
```
I noticed that there is an inconsistency here, where it is always called with 10 runs, but the default is set to 5. Regardless, even if this is an inconsistency - this is not really the place to change it and should be separate. I've reverted it back.

DanielRyanSmith commented on tools/wptrunner/wptrunner/stability.py:
```diff
@@ -346,38 +373,61 @@ def check_stability(logger, repeat_loop=10, repeat_restart=5, chaos_mode=True, m
         logger.info(':::')
         results, inconsistent, slow, iterations = step_func(**kwargs)
+
+        if iterations <= 1:
```
Good idea - changed!

DanielRyanSmith commented on tools/wptrunner/wptrunner/stability.py:
```diff
@@ -290,6 +298,12 @@ def wrap_handler(x):
+    # Use the number of repeated test suites that were run
+    # to process the results if the runs were stopped to
+    # avoid hitting the maximum run time.
+    if kwargs["avoided_timeout"]["did_avoid"]:
```
Yes, I was hesitant about taking this approach. I know we definitely want to have access to the number of actual iterations that were run, but returning that information from wptrunner's `run_tests` implementation in a circumstance that it is only used when called by stability tests didn't feel right either. I am totally open to changing this implementation if anyone has suggestions or thinks a rework of `run_tests` is worth the change.

jgraham commented on tools/wptrunner/wptrunner/wptrunner.py:
```diff
@@ -235,143 +360,83 @@ def run_tests(config, test_paths, product, **kwargs):
+                # if the next repeat run could cause the TC timeout to be
+                # reached, stop now and use the test results we have.
+                estimate = datetime.now() + longest_iteration_time
+                if not repeat_until_unexpected and max_time \
```
Please avoid using line continuation characters: it's preferred to use parens i.e.
```py
if (not repeat_until_unexpected and max_time
    and estimate >= start_time + max_time):
    # Condition body
```

DanielRyanSmith replied:
I see, I'll approach it like this in the future!

jgraham commented on tools/wptrunner/wptrunner/wptrunner.py:
```diff
@@ -235,143 +360,83 @@ def run_tests(config, test_paths, product, **kwargs):
+                if not repeat_until_unexpected and max_time \
+                        and estimate >= start_time + max_time:
+                    iteration_timeout = True
+                    logger.info(
```
Although there's one existing example of the "a problem occured!" style of error mssage, I think I'd prefer to avoid that if possible. Also at risk of being needlessly pedantic, the formatting here which breaks things into three lines when they could be two seems needlessly wasteful of vertical space.

DanielRyanSmith replied:
I will remove this output and change the formatting 👍

jgraham commented on tools/wptrunner/wptrunner/wptrunner.py:
```diff
@@ -235,143 +360,83 @@ def run_tests(config, test_paths, product, **kwargs):
+                    logger.info(
+                        "Repeat runs are in danger of reaching timeout!"
+                        " Quitting early.")
+                    logger.info(
```
This could definitely be one or two lines, and could use a f-string (since we're touching the code anyway).

jgraham commented on tools/wptrunner/wptrunner/wptrunner.py:
```diff
@@ -235,143 +360,83 @@ def run_tests(config, test_paths, product, **kwargs):
-                    test_groups = test_source_cls.tests_by_group(tests, **test_source_kwargs)
-                except Exception:
-                    logger.critical("Loading tests failed")
+                    logger.info(
```
I don't really know why this ended up getting reformatted, by my count the original line was 78 columns (I think we enforce 100 columns in pyflakes, but certainly more than this).

DanielRyanSmith replied:
Formatting fixed

jgraham commented on tools/wptrunner/wptrunner/wptrunner.py:
```diff
@@ -235,143 +360,83 @@ def run_tests(config, test_paths, product, **kwargs):
-                unexpected_pass_total += unexpected_pass_count
-                logger.info("Got %i unexpected results, with %i unexpected passes" %
-                            (unexpected_count, unexpected_pass_count))
+                logger.info(
```
Not sure why this got reformatted.

DanielRyanSmith replied:
This was probably my change when I reconfigured the code to use a counts dictionary. I'll make the formatting change.

jgraham commented on tools/wptrunner/wptrunner/wptrunner.py:
```diff
@@ -235,143 +360,83 @@ def run_tests(config, test_paths, product, **kwargs):
+                # Note this iteration's runtime and pad the total time
+                # by 10% to ensure ample time for the next iteration(s).
+                iteration_runtime = datetime.now() - iteration_start
+                iteration_runtime = timedelta(
```
Padding here seems weird. Why not just pad when we compare to the remaining time?

DanielRyanSmith replied:
Yes, I agree. I think it makes much more sense to store the actual runtime (as the variable states!) and pad the numbers only upon the comparison. Good change!

jgraham commented on tools/wptrunner/wptrunner/wptrunner.py:
```diff
@@ -150,6 +152,133 @@ def get_pause_after_test(test_loader, **kwargs):
+
+        browser_cls = product.get_browser_cls(test_type)
+
+        browser_kwargs = \
```
So a bunch of the reformatting here feels unnecessary, and this use of a continuation line in particular is not in keeping with the general project style.

DanielRyanSmith replied:
I'll remove all instances of these line continuations - noted to not use these in the future!

jgraham commented on tools/wptrunner/wptrunner/wptrunner.py:
```diff
@@ -150,6 +152,133 @@ def get_pause_after_test(test_loader, **kwargs):
+
+
+def evaluate_runs(counts, iteration_timeout, run_test_kwargs):
+    """Evaluates the test counts after the
```
No need to split this across multiple lines.

DanielRyanSmith replied:
switched to 1-liner 👍

jgraham commented on tools/wptrunner/wptrunner/wptrunner.py:
```diff
@@ -235,143 +360,83 @@ def run_tests(config, test_paths, product, **kwargs):
+                    break
 
-    return unexpected_total == 0
+    return evaluate_runs(counts, iteration_timeout, kwargs)
```
I think we should make `run_tests` return `counts` directly. Then `start` can discard the extra information and we can avoid putting the output in a kwarg.

DanielRyanSmith replied:
Yes, looking at this possible solution, it's a relatively simple fix that can be easily implemented and will not affect other tests using wptrunner. I made a change to only return `counts["repeat"]` to only give back the number of iterations of the test suite that were run, rather than all of `counts`. Do you think this is a good idea? Or would it be functionally better to return all of `counts` in the chance that other uses might arise and need further access to the other counts?

jgraham reviewed (changes requested):
Some general comments:
* If you're refactoring code it's really helpful to have that in a seperate commit first and then the actual changes in a commits, such that the commits are reviewable one-at-a-time. Of course that doesn't always work out, but in this case I think it could have done.
* It looks like some kind of automated code formatting was applied. In general we've relied on pyflakes and not adopted automatic formatting (we couldn't enforce it anyway, since the project is used across multiple repos, and asking them to all use a specific version of a specific formatter is implausible). Also we're not aiming for 80 column lines; breaking before 100 is preferred.

In terms of the actual change; I think it's mostly good, but @stephenmcgruer is right (as usual) that using the kwargs as in in/out parameter feels like a bad idea. Even though we do a reasonable amount of in-place manipulation of the kwargs before passing them to `run_tests`, they aren't currently used for output. I think it will work out OK to update the function signature to return the extra data we need, and just drop it in `start` so we maintain compatibility with tools using that entry point.

Thanks for working on this!

DanielRyanSmith committed:
wptrunner reformatting changes suggested by @jgraham

DanielRyanSmith committed:
functional changes to stability tests suggested by @jgraham

DanielRyanSmith committed:
flake8 changes "line break before binary operator"

DanielRyanSmith committed:
change run_tests to return counts object

DanielRyanSmith committed:
ensure run_tests return values are properly accessed

Now that wptrunner's run_tests returns more than 1 value, the return type will be a tuple for the older variables that expect only 1 value. We need to ensure that we pull the expected first value (boolean) out of that tuple.

DanielRyanSmith committed:
run_tests has consistent return values even in fail cases

wptrunner's run_tests would return a tuple only if not issues arose while running, and would return only a boolean in the case of some expected issue. Now a tuple is returned in all instances as expected.

DanielRyanSmith commented:
Thank you @stephenmcgruer and @jgraham for the reviews - I tried to remove some of the reformatting and I'll be sure to avoid any unnecessary auto-formatting changes and keep code refactoring in separate commits from now on.
Also, I would like to say that I appreciate how thorough the critiques of the code are, as it's helping me learn how to approach these issues moving forward!

jgraham reviewed (commented):
This is looking pretty reasonable now, but I think at least returning the whole of `counts` (or even a new `Status` class like a static version of `counts`) as the second argument is preferable to just returning the number of iterations, since it gives us a way to return richer data in the future without making breaking changes to the API.

DanielRyanSmith committed:
Return full counts in TestStatus class

run_tests will now return a new TestStatus object rather than returning only the number of iterations run. This will allow for more robust statistics to be shown in the future.

DanielRyanSmith committed:
small formatting changes

reducing some comments and logs to taking less vertical space.

DanielRyanSmith committed:
small wording change

TestStatus docstring

DanielRyanSmith commented:
@jgraham, I tried to implement as you suggested. Normally I would create a dataclass for this, but since I we need to support Python 3.6 and dataclasses were added in 3.7, I just created a class with attributes to return. Maybe there is a better way to handle this?

https://github.com/web-platform-tests/wpt/blob/9d27581be56b4280801d3e87c8fe2ff6b31a3ac1/tools/wptrunner/wptrunner/wptrunner.py#L273-L281

What would @jgraham do in this scenario? (I imagine I will be asking this often, if only to myself sometimes 😄 )

jgraham commented:
> What would @jgraham do in this scenario? 

I apologise in advance for any kind distress caused by thinking along those lines ;)

In this case I wonder if we could just have the `TestStatus` class and remove `counts` entirely, since we're not using the ability to add arbitary keys for anything other than convenience. So we'd change e.g. `counts["skipped"] += 1` to `status.skipped += 1`. I agree that using a dataclass will be nice once we can depend on 3.7+.

DanielRyanSmith committed:
Replace counts with TestStatus object

forego the use of defaultdict counts keeping track of test info/results and instead use the custom class TestStatus.

DanielRyanSmith committed:
convert some log strings to f-strings

jgraham reviewed (approved):
Thanks! I think this looks good now. If the CI passes feel free to squash and merge with an appropriate commit message. If the decision task fails again we'll need to investigate what's going on; it seems unlikely to be this specific changeset.
